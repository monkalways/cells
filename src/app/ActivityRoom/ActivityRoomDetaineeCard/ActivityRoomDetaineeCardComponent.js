import React from 'react';
import PropTypes from 'prop-types';
import { Card, withStyles } from '@material-ui/core';

import {
  CellDetaineeCardHeader,
  CellDetaineeCardContent,
} from '../../common/CellDetaineeCard';
import CardFooter from './CardFooter';

const propTypes = {
  classes: PropTypes.shape({}).isRequired,
  detainee: PropTypes.shape({
    firstName: PropTypes.string.isRequired,
    lastName: PropTypes.string.isRequired,
  }).isRequired,
  usage: PropTypes.string.isRequired,
  isAuthenticated: PropTypes.bool.isRequired,
  isCheckingIn: PropTypes.bool.isRequired,
  onCheckIn: PropTypes.func.isRequired,
};

const ActivityRoomDetaineeCardComponent = ({
  classes,
  detainee,
  usage,
  isAuthenticated,
  isCheckingIn,
  onCheckIn,
}) => (
  <Card className={classes.card}>
    <CellDetaineeCardHeader cellDetainee={detainee} />
    <CellDetaineeCardContent
      cellDetainee={detainee}
      first="activity-rooms"
      second={usage}
      isAuthenticated={isAuthenticated}
    />
    <CardFooter
      usage={usage}
      detainee={detainee}
      isAuthenticated={isAuthenticated}
      isCheckingIn={isCheckingIn}
      onCheckIn={onCheckIn}
    />
  </Card>
);

ActivityRoomDetaineeCardComponent.propTypes = propTypes;

export default withStyles((theme) => ({
  card: {
    maxWidth: theme.spacing.unit * 50,
    backgroundColor: theme.palette.background.paper,
  },
}))(ActivityRoomDetaineeCardComponent);
