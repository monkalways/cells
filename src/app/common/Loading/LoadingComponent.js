import React from 'react';
import PropTypes from 'prop-types';
import { CircularProgress, withStyles } from '@material-ui/core';

const propTypes = {
  classes: PropTypes.shape({}).isRequired,
  size: PropTypes.number,
};

const defaultProps = {
  size: 100,
};

const LoadingComponent = ({ classes, size }) => (
  <div className={classes.container}>
    <CircularProgress className={classes.progress} size={size} />
  </div>
);

LoadingComponent.propTypes = propTypes;
LoadingComponent.defaultProps = defaultProps;

export default withStyles((theme) => ({
  container: {
    display: 'flex',
    width: '100%',
    height: '100%',
    justifyContent: 'center',
    alignItems: 'center',
  },
  progress: {
    margin: theme.spacing.unit * 2,
  },
}))(LoadingComponent);
