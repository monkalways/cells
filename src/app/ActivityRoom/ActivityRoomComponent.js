import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Grid, withStyles, Typography } from '@material-ui/core';
import { withRouter } from 'react-router-dom';
import { compose } from 'recompose';

import ActivityRoomDetaineeCard from './ActivityRoomDetaineeCard';
import Header from './Header';
import Footer from './Footer';
import Layout from '../common/Layout';
import CellDetaineeGrid from '../common/CellDetaineeGrid';
import Loading from '../common/Loading';
import commonConstants from '../constants';

const propTypes = {
  classes: PropTypes.shape({}).isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      usage: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
  isActivityRoomDetaineesLoaded: PropTypes.bool.isRequired,
  isAuthenticated: PropTypes.bool.isRequired,
  detainees: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string.isRequired,
  })),
  isCheckingIn: PropTypes.bool.isRequired,
  isCheckingInSuccess: PropTypes.bool.isRequired,
  getActivityRoomDetainees: PropTypes.func.isRequired,
  handleCheckIn: PropTypes.func.isRequired,
  handleSignIn: PropTypes.func.isRequired,
  logOut: PropTypes.func.isRequired,
  refreshAuthenticationTimeout: PropTypes.func.isRequired,
  startAuthenticationTimeout: PropTypes.func.isRequired,
  stopAuthenticationTimeout: PropTypes.func.isRequired,
};

const defaultProps = {
  detainees: [],
};

export class ActivityRoomComponent extends Component {
  componentDidMount() {
    const { match, getActivityRoomDetainees } = this.props;
    const { usage } = match.params;
    getActivityRoomDetainees(usage);

    this.setUnauthenticatedTimeout();
  }

  componentDidUpdate() {
    const {
      isAuthenticated,
      logOut,
      match,
      startAuthenticationTimeout,
      stopAuthenticationTimeout,
    } = this.props;
    const { name } = match.params;
    if (isAuthenticated) {
      this.cancelUnauthenticatedTimeout();
      startAuthenticationTimeout(() => {
        stopAuthenticationTimeout();
        logOut('cells', name);
      });
    }
  }

  componentWillUnmount() {
    this.cancelUnauthenticatedTimeout();
  }

  setUnauthenticatedTimeout = () => {
    const { isAuthenticated } = this.props;
    this.cancelUnauthenticatedTimeout();
    if (!isAuthenticated) {
      this.unauthenticatedTimeout = setTimeout(() => {
        this.handleLogout();
      }, commonConstants.UNAUTHENTICATED_TIMEOUT_SECONDS * 1000);
    }
  };

  cancelUnauthenticatedTimeout = () => {
    if (this.unauthenticatedTimeout) {
      clearTimeout(this.unauthenticatedTimeout);
    }
  };

  handleLogout = () => {
    const { logOut, match } = this.props;
    const { usage } = match.params;
    logOut('activity-rooms', usage);
  };

  handleClick = () => {
    const { isAuthenticated, refreshAuthenticationTimeout } = this.props;
    if (isAuthenticated) {
      refreshAuthenticationTimeout();
    }
  };

  render() {
    const {
      match,
      detainees,
      classes,
      isCheckingIn,
      isCheckingInSuccess,
      isAuthenticated,
      isActivityRoomDetaineesLoaded,
      handleSignIn,
      handleCheckIn,
    } = this.props;
    const { usage } = match.params;
    return (
      <div
        onClick={() => this.handleClick()}
        role="presentation"
        id="refreshAuthenticationTimeoutHandler"
      >
        <React.Fragment>
          <Layout>
            <Header
              usage={usage}
              onLogout={this.handleLogout}
              isAuthenticated={isAuthenticated}
            />
            <div className={classes.body}>
              <CellDetaineeGrid>
                {isActivityRoomDetaineesLoaded ? (
                  detainees.length > 0 ? (
                    <React.Fragment>
                      {detainees.map((detainee) => (
                        <Grid key={detainee.id} item sm={4}>
                          <ActivityRoomDetaineeCard
                            detainee={detainee}
                            usage={usage}
                            isAuthenticated={isAuthenticated}
                            isCheckingIn={isCheckingIn}
                            isCheckingInSuccess={isCheckingInSuccess}
                            onCheckIn={handleCheckIn}
                          />
                        </Grid>
                      ))}
                    </React.Fragment>
                  ) : (
                    <Typography variant="h6" className={classes.heading}>
                      No detainees.
                    </Typography>
                  )
                ) : (
                  <Loading />
                )}
              </CellDetaineeGrid>
              <Footer
                isAuthenticated={isAuthenticated}
                onSignIn={handleSignIn}
              />
            </div>
          </Layout>
        </React.Fragment>
      </div>
    );
  }
}

ActivityRoomComponent.propTypes = propTypes;
ActivityRoomComponent.defaultProps = defaultProps;

export default compose(
  withStyles((theme) => ({
    body: {
      marginTop: theme.spacing.unit,
    },
    heading: {
      marginTop: theme.spacing.unit,
    },
  })),
  withRouter,
)(ActivityRoomComponent);
