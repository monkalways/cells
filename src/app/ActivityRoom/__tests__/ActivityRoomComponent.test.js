import React from 'react';
import { shallow } from 'enzyme';

import { Typography } from '@material-ui/core';
import { ActivityRoomComponent } from '../ActivityRoomComponent';
import Loading from '../../common/Loading';
import ActivityRoomDetaineeCard from '../ActivityRoomDetaineeCard';
import commonConstants from '../../constants';

describe('ActivityRoomComponent', () => {
  let push;
  let usage;
  let isActivityRoomDetaineesLoaded;
  let isAuthenticated;
  let detainees;
  let isCheckingIn;
  let isCheckingInSuccess;
  let getActivityRoomDetainees;
  let handleCheckIn;
  let handleSignIn;
  let handleLogOut;
  let refreshAuthenticationTimeout;
  let startAuthenticationTimeout;
  let stopAuthenticationTimeout;

  beforeEach(() => {
    push = jest.fn();
    usage = 'phone';
    isActivityRoomDetaineesLoaded = false;
    isAuthenticated = false;
    detainees = [];
    isCheckingIn = false;
    isCheckingInSuccess = false;
    getActivityRoomDetainees = jest.fn();
    handleCheckIn = jest.fn();
    handleSignIn = jest.fn();
    handleLogOut = jest.fn();
    refreshAuthenticationTimeout = jest.fn();
    startAuthenticationTimeout = jest.fn();
    stopAuthenticationTimeout = jest.fn();

    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  const setup = () => {
    const classes = {
      body: '',
      heading: '',
    };
    return shallow(<ActivityRoomComponent
      classes={classes}
      history={{ push }}
      match={{ params: { usage } }}
      isActivityRoomDetaineesLoaded={isActivityRoomDetaineesLoaded}
      detainees={detainees}
      isAuthenticated={isAuthenticated}
      isCheckingIn={isCheckingIn}
      isCheckingInSuccess={isCheckingInSuccess}
      getActivityRoomDetainees={getActivityRoomDetainees}
      handleCheckIn={handleCheckIn}
      handleSignIn={handleSignIn}
      handleLogOut={handleLogOut}
      refreshAuthenticationTimeout={refreshAuthenticationTimeout}
      startAuthenticationTimeout={startAuthenticationTimeout}
      stopAuthenticationTimeout={stopAuthenticationTimeout}
    />);
  };

  it('should render Loading when detainees are not yet loaded', () => {
    isActivityRoomDetaineesLoaded = false;
    const wrapper = setup();

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Loading)).toExist();
  });

  it('should render "no detainees" when detainees are loaded but there are none', () => {
    isActivityRoomDetaineesLoaded = true;
    detainees = [];
    const wrapper = setup();

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Typography)).toExist();
    expect(wrapper.find(Typography).html()).toContain('No detainees');
  });

  it('should render detainee cards when one or more detainees are loaded', () => {
    isActivityRoomDetaineesLoaded = true;
    detainees = [
      {
        id: '123',
      },
      {
        id: '456',
      },
    ];
    const wrapper = setup();

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(ActivityRoomDetaineeCard).length).toBe(2);
  });

  it('should get detainees when componentDidMount is called', () => {
    const wrapper = setup();

    wrapper.instance().componentDidMount();

    expect(getActivityRoomDetainees).toBeCalledWith(usage);
  });

  it('should not set unauthenticated timer when authenticated', () => {
    isAuthenticated = true;
    const wrapper = setup();

    wrapper.instance().componentDidMount();

    expect(setTimeout).not.toBeCalled();
  });

  it(`should set unauthenticated timer to call logout in ${
    commonConstants.UNAUTHENTICATED_TIMEOUT_SECONDS
  } seconds when not authenticated and clear timeout after unmount`, () => {
    isAuthenticated = false;
    const wrapper = setup();

    wrapper.instance().componentDidMount();
    expect(setTimeout).toBeCalled();

    jest.advanceTimersByTime(commonConstants.UNAUTHENTICATED_TIMEOUT_SECONDS * 1000);
    expect(handleLogOut).toBeCalledWith('activity-rooms', usage);

    wrapper.instance().componentWillUnmount();
    expect(clearTimeout).toBeCalled();
  });

  it('should cancel unauthenticated timer after authenticated is set to true after initialization', () => {
    isAuthenticated = false;
    const wrapper = setup();

    wrapper.instance().componentDidMount();
    expect(setTimeout).toBeCalled();

    wrapper.setProps({ isAuthenticated: true });
    expect(clearTimeout).toBeCalled();
  });

  it('should start authentication timeout if componentDidUpdate is called and page is newly authenticated', () => {
    isAuthenticated = false;
    const wrapper = setup();

    wrapper.setProps({ isAuthenticated: true });
    expect(startAuthenticationTimeout).toBeCalled();
  });

  it('should not start authentication timeout if componentDidUpdate is called and page is not authenticated', () => {
    isAuthenticated = false;
    const wrapper = setup();

    wrapper.setProps({ isAuthenticated: false });
    expect(startAuthenticationTimeout).not.toBeCalled();
  });

  it('should not start authentication timeout if componentDidUpdate is called and page was already authenticated', () => {
    isAuthenticated = false;
    const wrapper = setup();

    wrapper.setProps({ isAuthenticated: true });
    wrapper.setProps({ isAuthenticated: true });
    expect(startAuthenticationTimeout).toBeCalledTimes(1);
  });

  it('should refresh authentication timeout on handleClick if page is authenticated', () => {
    isAuthenticated = true;
    const wrapper = setup();

    wrapper.find('#authenticatedActivityRoomRoot').simulate('click');
    expect(refreshAuthenticationTimeout).toBeCalled();
  });

  it('should not refresh authentication timeout on handleClick if page is not authenticated', () => {
    isAuthenticated = false;
    const wrapper = setup();

    wrapper.find('#authenticatedActivityRoomRoot').simulate('click');
    expect(refreshAuthenticationTimeout).not.toBeCalled();
  });
});
