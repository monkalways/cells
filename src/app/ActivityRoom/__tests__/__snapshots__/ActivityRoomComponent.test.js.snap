// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityRoomComponent should render "no detainees" when detainees are loaded but there are none 1`] = `
<div
  id="refreshAuthenticationTimeoutHandler"
  onClick={[Function]}
  role="presentation"
>
  <WithStyles(LayoutComponent)>
    <WithStyles(HeaderComponent)
      isAuthenticated={false}
      onLogout={[Function]}
      usage="phone"
    />
    <div
      className=""
    >
      <WithStyles(CellDetaineeGridComponent)>
        <WithStyles(Typography)
          className=""
          variant="h6"
        >
          No detainees.
        </WithStyles(Typography)>
      </WithStyles(CellDetaineeGridComponent)>
      <WithStyles(withRouter(FooterComponent))
        isAuthenticated={false}
        onSignIn={[MockFunction]}
      />
    </div>
  </WithStyles(LayoutComponent)>
</div>
`;

exports[`ActivityRoomComponent should render Loading when detainees are not yet loaded 1`] = `
<div
  id="refreshAuthenticationTimeoutHandler"
  onClick={[Function]}
  role="presentation"
>
  <WithStyles(LayoutComponent)>
    <WithStyles(HeaderComponent)
      isAuthenticated={false}
      onLogout={[Function]}
      usage="phone"
    />
    <div
      className=""
    >
      <WithStyles(CellDetaineeGridComponent)>
        <WithStyles(LoadingComponent) />
      </WithStyles(CellDetaineeGridComponent)>
      <WithStyles(withRouter(FooterComponent))
        isAuthenticated={false}
        onSignIn={[MockFunction]}
      />
    </div>
  </WithStyles(LayoutComponent)>
</div>
`;

exports[`ActivityRoomComponent should render detainee cards when one or more detainees are loaded 1`] = `
<div
  id="refreshAuthenticationTimeoutHandler"
  onClick={[Function]}
  role="presentation"
>
  <WithStyles(LayoutComponent)>
    <WithStyles(HeaderComponent)
      isAuthenticated={false}
      onLogout={[Function]}
      usage="phone"
    />
    <div
      className=""
    >
      <WithStyles(CellDetaineeGridComponent)>
        <WithStyles(Grid)
          item={true}
          key="123"
          sm={4}
        >
          <WithStyles(ActivityRoomDetaineeCardComponent)
            detainee={
              Object {
                "id": "123",
              }
            }
            isAuthenticated={false}
            isCheckingIn={false}
            isCheckingInSuccess={false}
            onCheckIn={[MockFunction]}
            usage="phone"
          />
        </WithStyles(Grid)>
        <WithStyles(Grid)
          item={true}
          key="456"
          sm={4}
        >
          <WithStyles(ActivityRoomDetaineeCardComponent)
            detainee={
              Object {
                "id": "456",
              }
            }
            isAuthenticated={false}
            isCheckingIn={false}
            isCheckingInSuccess={false}
            onCheckIn={[MockFunction]}
            usage="phone"
          />
        </WithStyles(Grid)>
      </WithStyles(CellDetaineeGridComponent)>
      <WithStyles(withRouter(FooterComponent))
        isAuthenticated={false}
        onSignIn={[MockFunction]}
      />
    </div>
  </WithStyles(LayoutComponent)>
</div>
`;
